docs/RFCS/20170920_sql_consistency_checks.md

```diff
@@ -0,0 +1,428 @@
+- Invalid data encodings
+- Non-canonical encodings (where we can decode the data, but if we
+  re-encode it we get a different value.)
+- [Composite encodings][] are valid. This includes `DECIMAL` and
+  collated strings types
+- Key encoding correctly reflects the ordering of the decoded values
+
+Schemas also require checks for the following:
+- Invalid parent ID for interleaved tables
+- Invalid index definitions (wrong columns)
+- Invalid view descriptors: invalid column types; invalid SQL; invalid
+  number of columns.
+- Invalid column families
+
+Outside of user data, we can also check system tables.
+TODO: What checks are relevant here?
+
+By having consistency checks run periodically we can understand if and
+when errors are occurring. As a user of cockroach, I will know my data
+is correct and safe. Furthermore, these checks will be an asset for
+running during tests. For example, they can be run after each acceptance
+and logic test.
+
+[#17626]: https://github.com/cockroachdb/cockroach/issues/17626
+[#17690]: https://github.com/cockroachdb/cockroach/issues/17690
+[#18533]: https://github.com/cockroachdb/cockroach/issues/18533
+[#18705]: https://github.com/cockroachdb/cockroach/issues/18705
+[STORING]: https://www.cockroachlabs.com/docs/stable/create-index.html#store-columns
+[FAMILY]: https://www.cockroachlabs.com/docs/stable/column-families.html
+[Composite encodings]: https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/encoding.md#composite-encoding
+
+# Guide-level explanation
+
+## Terminology
+
+The **physical table** is the underlying KV pairs behind tables. This is
+where we find primary and secondary indexes (as distinct keys),
+groupings of certain data (column families), how index data is organized
+(index storing conditions), and the encoding of the data.
+
+On the other hand, the **logical table** is the SQL representation of
+the data. This is where we think about data as rows and columns. It's
+also where we have table schemas and their corresponding constraints
+(`UNIQUE`, `CHECK`, ...).
+
+Checks fall into two categories for their types. They are either a check
+on the physical table or the logical table. For example, a grouping of
+data only exist as the specific value encoded in the KV pairs
+corresponding to an SQL row, and as such the only to can access them is
+through the KV layer. Secondary indexes are an exception to this strict
+bucketing as they exist in physical tables but have been intentionally
+exposed through SQL interfaces.
+
+Alongside the check type, there are two main types of implementations of
+the checks that will be discussed in this RFC. The first is using SQL
+queries to execute the check, which can be used to check the logical
+table[0] -- hence we will call them **SQL checks**. As for the physical
+layer checks, these need to be done through operating on the KV pairs
+directly -- so we will call these **KV checks**. The method for
+running KV checks in 
```

Comment:
It does look like we want a single process that is running all the checks (or all configured checks) rather than separate processors running different checks. Which points to not need to separate the checking into kv checks and sql checks. I found the discussion below quite difficult to read. What I hoping to read about was a presentation of a SQL command that would validate a table by running all the validation checks on it together.